examples/integrations/hf_trainer_api/det_callback.py

```diff
@@ -0,0 +1,348 @@
+import pathlib
+import typing
+
+from transformers import (
+    TrainerCallback,
+    TrainerState,
+    TrainerControl,
+    TrainingArguments,
+)
+
+import os
+import determined as det
+import logging
+import torch
+import uuid
+import weakref
+import importlib.util
+import shutil
+
+
+class DetCallback(TrainerCallback):
+    def __init__(
+        self,
+        args: TrainingArguments,
+        filter_metrics: typing.List[str] = None,
+        tokenizer: typing.Any = None,
+        tokenizer_options: typing.Dict = None,
+        checkpoint_metadata: typing.Dict = None,
+    ) -> None:
+        super().__init__()
+
+        assert (
+            is_determined_available()
+        ), "DetCallback requires determined to be installed. Run `pip install determined`."
+        import determined
+
+        self._det = determined
+        if args.deepspeed:
+            distributed = self._det.core.DistributedContext.from_deepspeed()
+        else:
+            distributed = self._det.core.DistributedContext.from_torch_distributed()
+        self.core_context = self._det.core.init(distributed=distributed)
+        self.core_context.__enter__()
+        weakref.finalize(self, exit_context, self.core_context)
```

Comment:
I think this is a bad idea.

It _does_ guarantee that core_context.close() is called, eventually.  Though if it happens during interpreter shutdown you can get some crazy confusing stacktraces (which we've seen before](https://determinedai.atlassian.net/browse/DET-2302)).

But there are other useful things that core_context does, like catching InvalidHP exceptions.  And some day, I'd like to add asynchronous metrics and/or profiling and/or tensorboard logging, in which case we'll need a finish() or flush() call.  And that finish() or flush() call will not be safe to try to run during interpreter shutdown.

I think a cleaner API would be to require the user to do something like:

```python

def my_main(core_context):
    trainer.Train(..., callbacks=[DetCallback(core_context)])

with det.core.init() as core_context:
    my_main(core_context)
```